php_phongo.c

```diff
@@ -2503,6 +2507,64 @@ int php_phongo_set_monitoring_callbacks(mongoc_client_t* client)
 	return retval;
 }
 
+static zval* php_phongo_manager_prepare_manager_for_hash(zval* driverOptions, bool* free TSRMLS_DC)
+{
+	php_phongo_manager_t* manager;
+	zval*                 autoEncryptionOpts      = NULL;
+	zval*                 keyVaultClient          = NULL;
+	zval*                 driverOptionsClone      = NULL;
+	zval*                 autoEncryptionOptsClone = NULL;
+#if PHP_VERSION_ID >= 70000
+	zval stackAutoEncryptionOptsClone;
+#endif
+
+	if (!driverOptions) {
+		return NULL;
+	}
+
+	if (!php_array_existsc(driverOptions, "autoEncryption")) {
+		goto ref;
+	}
+
+	autoEncryptionOpts = php_array_fetchc(driverOptions, "autoEncryption");
+	if (Z_TYPE_P(autoEncryptionOpts) != IS_ARRAY) {
+		goto ref;
+	}
+
+	if (!php_array_existsc(autoEncryptionOpts, "keyVaultClient")) {
+		goto ref;
+	}
+
+	keyVaultClient = php_array_fetchc(autoEncryptionOpts, "keyVaultClient");
+	if (Z_TYPE_P(keyVaultClient) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(keyVaultClient), php_phongo_manager_ce TSRMLS_CC)) {
+		goto ref;
+	}
+
+	*free = true;
+
+	manager = Z_MANAGER_OBJ_P(keyVaultClient);
+
+#if PHP_VERSION_ID >= 70000
+	driverOptionsClone = ecalloc(sizeof(zval), 1);
+	autoEncryptionOptsClone = &stackAutoEncryptionOptsClone;
+#else
+	ALLOC_INIT_ZVAL(driverOptionsClone);
+	MAKE_STD_ZVAL(autoEncryptionOptsClone);
```

Comment:
@jmikola: Either this or the previous line cause a significant number of leaks on PHP 5. I'll investigate this separately once I manage to get PHP 5.6 running on my machine.

```
[2019/12/06 15:51:52.924] 051+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 052+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_hash.c(320) :  Freeing 0x7F81A23D3070 (64 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 053+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_alloc.c(2592) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 054+ Last leak repeated 21 times
[2019/12/06 15:51:52.924] 055+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 056+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_hash.c(346) :  Freeing 0x7F81A23D3108 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 057+ Last leak repeated 25 times
[2019/12/06 15:51:52.924] 058+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 059+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(4077) :  Freeing 0x7F81A23D31A8 (32 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 060+ Last leak repeated 7 times
[2019/12/06 15:51:52.924] 061+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 062+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(9481) :  Freeing 0x7F81A23D32C0 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 063+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_API.c(1011) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 064+ Last leak repeated 9 times
[2019/12/06 15:51:52.924] 065+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 066+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(9414) :  Freeing 0x7F81A23D3360 (32 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 067+ Last leak repeated 16 times
[2019/12/06 15:51:52.924] 068+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 069+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(11818) :  Freeing 0x7F81A23D3510 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 070+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_API.c(1011) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 071+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 072+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(11751) :  Freeing 0x7F81A23D35B0 (32 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 073+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 074+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_hash.c(392) :  Freeing 0x7F81A23D3628 (64 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 075+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_alloc.c(2592) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 076+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 077+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(11802) :  Freeing 0x7F81A23D36C0 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 078+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_hash.c(419) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 079+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 080+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(4138) :  Freeing 0x7F81A23D4BF0 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 081+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_API.c(1011) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 082+ Last leak repeated 3 times
[2019/12/06 15:51:52.924] 083+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 084+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_API.c(314) :  Freeing 0x7F81A23D5DE8 (32 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 085+ Last leak repeated 13 times
[2019/12/06 15:51:52.924] 086+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 087+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(29216) :  Freeing 0x7F81A25A06D0 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 088+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_API.c(1011) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 089+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 090+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(944) :  Freeing 0x7F81A25A28D8 (32 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 091+ [Fri Dec  6 20:51:52 2019]  Script:  '/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php'
[2019/12/06 15:51:52.924] 092+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_vm_execute.h(16156) :  Freeing 0x7F81A25A2A40 (72 bytes), script=/data/mci/0bd12b2c9b9c1e72c0226b4a04828642/src/tests/manager/manager-ctor-auto_encryption-001.php
[2019/12/06 15:51:52.924] 093+ /data/mci/0f02f769a25a722af572a3a28e934dc8/src/php-5.6.38/Zend/zend_API.c(1011) : Actual location (location was relayed)
[2019/12/06 15:51:52.924] 094+ === Total 108 memory leaks detected ===
```

Reply:
Fixed - a tiny `Z_ADDREF_P` slipped through the cracks and was having a field day in memory. It is no more.